>  so there must be something in your changes that cause it

I don't see how that's possible.   If I check out main, git pull --rebase, and run the tests, they fail.  My code is all in a different branch.

```
(airflow-env) ferruzzi:~/workplace/airflow (ferruzzi/boto-user-agent)
$ git checkout main
Switched to branch 'main'
Your branch is up to date with 'apache/main'.
(airflow-env) ferruzzi:~/workplace/airflow (main)
$ git pull --rebase
Already up to date.
Current branch main is up to date.
(airflow-env) ferruzzi:~/workplace/airflow (main)
$ pytest tests/providers/amazon/aws/hooks/test_s3.py::TestAwsS3HookNoMock::test_check_for_bucket_raises_error_with_invalid_conn_id -W ignore::DeprecationWarning -W  ignore::FutureWarning
=========================================================================== test session starts ===========================================================================
platform linux -- Python 3.8.10, pytest-6.2.5, py-1.11.0, pluggy-1.0.0 -- /home/ANT.AMAZON.COM/ferruzzi/.pyenv/versions/airflow-env/bin/python
cachedir: .pytest_cache
rootdir: /home/ANT.AMAZON.COM/ferruzzi/workplace/airflow, configfile: pytest.ini
plugins: anyio-3.6.2, xdist-2.5.0, forked-1.4.0, requests-mock-1.9.3, flaky-3.7.0, instafail-0.4.2, rerunfailures-9.1.1, cov-3.0.0, timeouts-1.2.1, httpx-0.15.0, asyncio-0.16.0
setup timeout: 0.0s, execution timeout: 0.0s, teardown timeout: 0.0s
collected 1 item                                                                                                                                                          

tests/providers/amazon/aws/hooks/test_s3.py::TestAwsS3HookNoMock::test_check_for_bucket_raises_error_with_invalid_conn_id FAILED                                    [100%]

================================================================================ FAILURES =================================================================================
_______________________________________________ TestAwsS3HookNoMock.test_check_for_bucket_raises_error_with_invalid_conn_id _______________________________________________

self = <tests.providers.amazon.aws.hooks.test_s3.TestAwsS3HookNoMock object at 0x7f7d4a39c610>, monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7f7d4a39ca30>

    def test_check_for_bucket_raises_error_with_invalid_conn_id(self, monkeypatch):
        monkeypatch.delenv("AWS_PROFILE", raising=False)
        monkeypatch.delenv("AWS_ACCESS_KEY_ID", raising=False)
        monkeypatch.delenv("AWS_SECRET_ACCESS_KEY", raising=False)
        hook = S3Hook(aws_conn_id="does_not_exist")
        # We're mocking all actual AWS calls and don't need a connection. This
        # avoids an Airflow warning about connection cannot be found.
        hook.get_connection = lambda _: None
        with pytest.raises(NoCredentialsError):
>           hook.check_for_bucket("test-non-existing-bucket")
E           Failed: DID NOT RAISE <class 'botocore.exceptions.NoCredentialsError'>

tests/providers/amazon/aws/hooks/test_s3.py:63: Failed
-------------------------------------------------------------------------- Captured stdout setup --------------------------------------------------------------------------
========================= AIRFLOW ==========================
Home of the user: /home/ANT.AMAZON.COM/ferruzzi
Airflow home /home/ANT.AMAZON.COM/ferruzzi/airflow
Skipping initializing of the DB as it was initialized already.
You can re-initialize the database by adding --with-db-init flag when running tests.
-------------------------------------------------------------------------- Captured stdout call ---------------------------------------------------------------------------
[2022-12-03 10:47:34,076] {base_aws.py:121} INFO - No connection ID provided. Fallback on boto3 credential strategy (region_name=None). See: https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html
[2022-12-03 10:47:34,087] {credentials.py:1251} INFO - Found credentials in shared credentials file: ~/.aws/credentials
[2022-12-03 10:47:34,593] {s3.py:221} INFO - Bucket "test-non-existing-bucket" does not exist
---------------------------------------------------------------------------- Captured log call ----------------------------------------------------------------------------
INFO     airflow.providers.amazon.aws.hooks.base_aws.BaseSessionFactory:base_aws.py:121 No connection ID provided. Fallback on boto3 credential strategy (region_name=None). See: https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html
INFO     botocore.credentials:credentials.py:1251 Found credentials in shared credentials file: ~/.aws/credentials
INFO     airflow.providers.amazon.aws.hooks.s3.S3Hook:s3.py:221 Bucket "test-non-existing-bucket" does not exist
============================================================================ warnings summary =============================================================================
../../.pyenv/versions/airflow-env/lib/python3.8/site-packages/_pytest/config/__init__.py:1233
  /home/ANT.AMAZON.COM/ferruzzi/.pyenv/versions/airflow-env/lib/python3.8/site-packages/_pytest/config/__init__.py:1233: PytestConfigWarning: Unknown config option: asyncio_mode
  
    self._warn_or_fail_if_strict(f"Unknown config option: {key}\n")

-- Docs: https://docs.pytest.org/en/stable/warnings.html
========================================================================= short test summary info =========================================================================
FAILED tests/providers/amazon/aws/hooks/test_s3.py::TestAwsS3HookNoMock::test_check_for_bucket_raises_error_with_invalid_conn_id - Failed: DID NOT RAISE <class 'botocor...
====================================================================== 1 failed, 1 warning in 1.73s =======================================================================
(airflow-env) ferruzzi:~/workplace/airflow (main)
$ 
```

--- a/airflow/providers/amazon/aws/hooks/base_aws.py
+++ b/airflow/providers/amazon/aws/hooks/base_aws.py
@@ -28,6 +28,7 @@ import datetime
 import json
 import logging
 import warnings
+from copy import deepcopy
 from functools import wraps
 from typing import TYPE_CHECKING, Any, Callable, Generic, TypeVar, Union
 
@@ -42,12 +43,14 @@ from botocore.credentials import ReadOnlyCredentials
 from dateutil.tz import tzlocal
 from slugify import slugify
 
+from airflow import __version__ as airflow_version
 from airflow.compat.functools import cached_property
 from airflow.configuration import conf
 from airflow.exceptions import AirflowException, AirflowNotFoundException
 from airflow.hooks.base import BaseHook
 from airflow.providers.amazon.aws.utils.connection_wrapper import AwsConnectionWrapper
 from airflow.utils.helpers import exactly_one
+from airflow.providers_manager import ProvidersManager
 from airflow.utils.log.logging_mixin import LoggingMixin
 from airflow.utils.log.secrets_masker import mask_secret
 
@@ -406,8 +409,23 @@ class AwsGenericHook(BaseHook, Generic[BaseAwsConnection]):
         self.resource_type = resource_type
 
         self._region_name = region_name
-        self._config = config
+        self._config = config or botocore.config.Config()
         self._verify = verify
+
+    @classmethod
+    def get_provider_version(cls) -> str:
+        manager = ProvidersManager()
+        provider_name = manager.hooks[cls.conn_type].package_name  # type: ignore[union-attr]
+        provider = manager.providers[provider_name]
+        return provider.version
+
+    def generate_user_agent_extra_field(self, existing_user_agent_extra: str) -> str:
+        user_agent_extra_values = [
+            f"Airflow/{airflow_version}",
+            f"AmPP/{self.get_provider_version()}",
+            existing_user_agent_extra or "",
+        ]
+        return " ".join(user_agent_extra_values).strip()
 
     @cached_property
     def conn_config(self) -> AwsConnectionWrapper:
@@ -451,6 +469,23 @@ class AwsGenericHook(BaseHook, Generic[BaseAwsConnection]):
             conn=self.conn_config, region_name=region_name, config=self.config
         ).create_session()
 
+    def _get_config(self, config: Config | None = None) -> Config:
+        """
+        No AWS Operators use the config argument to this method.
+        Keep backward compatibility with other users who might use it
+        """
+        if config is None:
+            config = deepcopy(self.config)
+
+        # ignore[union-attr] is required for this block to appease MyPy
+        # because the user_agent_extra field is generated at runtime.
+        user_agent_config = Config(
+            user_agent_extra=self.generate_user_agent_extra_field(
+                existing_user_agent_extra=config.user_agent_extra  # type: ignore[union-attr]
+            )
+        )
+        return config.merge(user_agent_config)  # type: ignore[union-attr]
+
     def get_client_type(
         self,
         region_name: str | None = None,
@@ -458,15 +493,12 @@ class AwsGenericHook(BaseHook, Generic[BaseAwsConnection]):
     ) -> boto3.client:
         """Get the underlying boto3 client using boto3 session"""
         client_type = self.client_type
-
-        # No AWS Operators use the config argument to this method.
-        # Keep backward compatibility with other users who might use it
-        if config is None:
-            config = self.config
-
         session = self.get_session(region_name=region_name)
         return session.client(
-            client_type, endpoint_url=self.conn_config.endpoint_url, config=config, verify=self.verify
+            client_type,
+            endpoint_url=self.conn_config.endpoint_url,
+            config=self._get_config(config),
+            verify=self.verify,
         )
 
     def get_resource_type(
@@ -476,15 +508,12 @@ class AwsGenericHook(BaseHook, Generic[BaseAwsConnection]):
     ) -> boto3.resource:
         """Get the underlying boto3 resource using boto3 session"""
         resource_type = self.resource_type
-
-        # No AWS Operators use the config argument to this method.
-        # Keep backward compatibility with other users who might use it
-        if config is None:
-            config = self.config
-
         session = self.get_session(region_name=region_name)
         return session.resource(
-            resource_type, endpoint_url=self.conn_config.endpoint_url, config=config, verify=self.verify
+            resource_type,
+            endpoint_url=self.conn_config.endpoint_url,
+            config=self._get_config(config),
+            verify=self.verify,
         )
 
     @cached_property

--- a/airflow/providers/amazon/aws/hooks/base_aws.py
+++ b/airflow/providers/amazon/aws/hooks/base_aws.py
@@ -420,6 +420,11 @@ class AwsGenericHook(BaseHook, Generic[BaseAwsConnection]):
         try:
             manager = ProvidersManager()
             hook = manager.hooks[cls.conn_type]
+            if not hook:
+                # This gets caught immediately, but without it MyPy complains
+                # Item "None" of "Optional[HookInfo]" has no attribute "package_name"
+                # on the following line and static checks fail.
+                raise ValueError(f"Hook info for {cls.conn_type} not found in the Provider Manager.")
             provider = manager.providers[hook.package_name]
             return provider.version
         except Exception:
@@ -466,7 +471,7 @@ class AwsGenericHook(BaseHook, Generic[BaseAwsConnection]):
         extracted from the resulting UUID.
         """
         try:
-            dag_id = os.getenv("AIRFLOW_CTX_DAG_ID")
+            dag_id = os.environ["AIRFLOW_CTX_DAG_ID"]
             return str(uuid.uuid5(uuid.NAMESPACE_OID, dag_id))
         except Exception:
             # Under no condition should an error here ever cause an issue for the user.
@@ -479,6 +484,7 @@ class AwsGenericHook(BaseHook, Generic[BaseAwsConnection]):
             # This can be a circular import under specific configurations.
             # Importing locally to either avoid or catch it if it does happen.
             from airflow import __version__ as airflow_version
+
             return airflow_version
         except Exception:
             # Under no condition should an error here ever cause an issue for the user.

--- a/airflow/providers/amazon/aws/hooks/base_aws.py
+++ b/airflow/providers/amazon/aws/hooks/base_aws.py
@@ -25,6 +25,7 @@ This module contains Base AWS Hook.
 from __future__ import annotations
 
 import datetime
+import inspect
 import json
 import logging
 import warnings
@@ -413,16 +414,50 @@ class AwsGenericHook(BaseHook, Generic[BaseAwsConnection]):
         self._verify = verify
 
     @classmethod
-    def get_provider_version(cls) -> str:
+    def _get_provider_version(cls) -> str:
+        """Checks the Providers Manager for the package version."""
         manager = ProvidersManager()
         provider_name = manager.hooks[cls.conn_type].package_name  # type: ignore[union-attr]
         provider = manager.providers[provider_name]
         return provider.version
 
-    def generate_user_agent_extra_field(self, existing_user_agent_extra: str) -> str:
+    @staticmethod
+    def _find_class_name(target_function_name: str) -> str:
+        """
+        Given a frame off the stack, return the name of the class which made the call.
+        Note: This method may raise a ValueError or an IndexError, but the calling
+        method is catching and handling those.
+        """
+        stack = inspect.stack()
+        # Find the index of the most recent frame which called the provided function name.
+        target_frame_index = [frame.function for frame in stack].index(target_function_name)
+        # Pull that frame off the stack.
+        target_frame = stack[target_frame_index][0]
+        # Get the local variables for that frame.
+        frame_variables = target_frame.f_locals["self"]
+        # Get the class object for that frame.
+        frame_class_object = frame_variables.__class__
+        # Return the name of the class object.
+        return frame_class_object.__name__
+
+    def _get_caller(self, target_function_name: str = "execute") -> str:
+        """Given a function name, walk the stack and return the name of the class which called it last."""
+        try:
+            caller = self._find_class_name(target_function_name)
+            if caller == "BaseSensorOperator":
+                # If the result is a BaseSensorOperator, then look for whatever last called "poke".
+                return self._get_caller("poke")
+            return caller
+        except Exception:
+            # While it is generally considered poor form to catch "Exception", under
+            # no condition should an error here ever cause an issue for the user.
+            return "Unknown"
+
+    def _generate_user_agent_extra_field(self, existing_user_agent_extra: str) -> str:
         user_agent_extra_values = [
             f"Airflow/{airflow_version}",
-            f"AmPP/{self.get_provider_version()}",
+            f"AmPP/{self._get_provider_version()}",
+            f"Caller/{self._get_caller()}",
             existing_user_agent_extra or "",
         ]
         return " ".join(user_agent_extra_values).strip()
@@ -480,7 +515,7 @@ class AwsGenericHook(BaseHook, Generic[BaseAwsConnection]):
         # ignore[union-attr] is required for this block to appease MyPy
         # because the user_agent_extra field is generated at runtime.
         user_agent_config = Config(
-            user_agent_extra=self.generate_user_agent_extra_field(
+            user_agent_extra=self._generate_user_agent_extra_field(
                 existing_user_agent_extra=config.user_agent_extra  # type: ignore[union-attr]
             )
         )

--- a/tests/providers/amazon/aws/hooks/test_base_aws.py
+++ b/tests/providers/amazon/aws/hooks/test_base_aws.py
@@ -301,6 +301,26 @@ class TestAwsBaseHook:
 
         assert table.item_count == 0
 
+    @pytest.mark.parametrize(
+        "client_meta",
+        [
+            AwsBaseHook(client_type="s3").get_client_type().meta,
+            AwsBaseHook(resource_type="dynamodb").get_resource_type().meta.client.meta,
+        ],
+    )
+    def test_user_agent_extra_update(self, client_meta):
+        """
+        We are only looking for the keys appended by the AwsBaseHook. A user_agent string
+        is a number of key/value pairs such as: `BOTO3/1.25.4 AIRFLOW/2.5.0.DEV0 AMPP/6.0.0`.
+        """
+        expected_user_agent_tag_keys = ["Airflow", "AmPP"]
+
+        result_user_agent_tags = client_meta.config.user_agent.split(" ")
+        result_user_agent_tag_keys = [tag.split("/")[0].lower() for tag in result_user_agent_tags]
+
+        for key in expected_user_agent_tag_keys:
+            assert key.lower() in result_user_agent_tag_keys
+
     @mock.patch.object(AwsBaseHook, "get_connection")
     @mock_sts
     def test_assume_role(self, mock_get_connection):
@@ -346,7 +366,7 @@ class TestAwsBaseHook:
             hook.get_client_type("s3")
 
         calls_assume_role = [
-            mock.call.session.Session().client("sts", config=None),
+            mock.call.session.Session().client("sts", config=mock.ANY),
             mock.call.session.Session()
             .client()
             .assume_role(
@@ -510,7 +530,7 @@ class TestAwsBaseHook:
             mock_xpath.assert_called_once_with(xpath)
 
         calls_assume_role_with_saml = [
-            mock.call.session.Session().client("sts", config=None),
+            mock.call.session.Session().client("sts", config=mock.ANY),
             mock.call.session.Session()
             .client()
             .assume_role_with_saml(
@@ -735,7 +755,7 @@ class TestAwsBaseHook:
         mock_session_factory.assert_called_once_with(
             conn=hook.conn_config,
             region_name=method_region_name,
-            config=hook_botocore_config,
+            config=mock.ANY,
         )
         assert mock_session_factory_instance.create_session.assert_called_once
         assert session == MOCK_BOTO3_SESSION
@@ -807,7 +827,7 @@ class ThrowErrorUntilCount:
 
     def __call__(self):
         """
-        Raise an Forbidden until after count threshold has been crossed.
+        Raise an Exception until after count threshold has been crossed.
         Then return True.
         """
         if self.counter < self.count:
@@ -839,7 +859,8 @@ class TestRetryDecorator:  # ptlint: disable=invalid-name
         result = _retryable_test(lambda: 42)
         assert result, 42
 
-    def test_retry_on_exception(self):
+    @mock.patch("time.sleep", return_value=0)
+    def test_retry_on_exception(self, _):
         quota_retry = {
             "stop_after_delay": 2,
             "multiplier": 1,

--- a/tests/providers/amazon/aws/hooks/test_base_aws.py
+++ b/tests/providers/amazon/aws/hooks/test_base_aws.py
@@ -313,13 +313,37 @@ class TestAwsBaseHook:
         We are only looking for the keys appended by the AwsBaseHook. A user_agent string
         is a number of key/value pairs such as: `BOTO3/1.25.4 AIRFLOW/2.5.0.DEV0 AMPP/6.0.0`.
         """
-        expected_user_agent_tag_keys = ["Airflow", "AmPP"]
+        expected_user_agent_tag_keys = ["Airflow", "AmPP", "Caller"]
 
         result_user_agent_tags = client_meta.config.user_agent.split(" ")
         result_user_agent_tag_keys = [tag.split("/")[0].lower() for tag in result_user_agent_tags]
 
         for key in expected_user_agent_tag_keys:
             assert key.lower() in result_user_agent_tag_keys
+
+    @staticmethod
+    def fetch_tags() -> dict[str:str]:
+        """Helper method which creates an AwsBaseHook and returns the user agent string split into a dict."""
+        user_agent_string = AwsBaseHook(client_type="s3").get_client_type().meta.config.user_agent
+        # Split the list of {Key}/{Value} into a dict
+        return dict(tag.split("/") for tag in user_agent_string.split(" "))
+
+    @pytest.mark.parametrize("found_classes", [["RandomOperator"], ["BaseSensorOperator", "TestSensor"]])
+    @mock.patch.object(AwsBaseHook, "_find_class_name")
+    def test_user_agent_caller_target_function_found(self, mock_class_name, found_classes):
+        mock_class_name.side_effect = found_classes
+
+        user_agent_tags = self.fetch_tags()
+
+        assert mock_class_name.call_count == len(found_classes)
+        assert user_agent_tags["Caller"] == found_classes[-1]
+
+    def test_user_agent_caller_target_function_not_found(self):
+        default_caller_name = "Unknown"
+
+        user_agent_tags = self.fetch_tags()
+
+        assert user_agent_tags["Caller"] == default_caller_name
 
     @mock.patch.object(AwsBaseHook, "get_connection")
     @mock_sts

